Deferred/FutureType.swift

```diff
@@ -0,0 +1,164 @@
+    /// value.
+    ///
+    /// - parameter time: A length of time to wait for the value to be determined.
+    /// - returns: The determined value, if filled within the timeout, or `nil`.
+    func wait(time: Timeout) -> Value?
+}
+
+public extension FutureType {
+    /// Call some function in the background once the value is determined.
+    ///
+    /// If the value is determined, the function will be dispatched immediately.
+    /// An `upon` call should always execute asynchronously.
+    ///
+    /// - parameter body: A function that uses the determined value.
+    func upon(body: Value -> ()) {
+        upon(genericQueue, body: body)
+    }
+
+    /// Call some function on the main queue once the value is determined.
+    ///
+    /// If the value is determined, the function will be submitted to the
+    /// main queue immediately. The function should always be executed
+    /// asynchronously, even if this function is called from the main queue.
+    ///
+    /// - parameter body: A function that uses the determined value.
+    func uponMainQueue(body: Value -> ()) {
+        upon(dispatch_get_main_queue(), body: body)
+    }
+}
+
+public extension FutureType {
+    /// Checks for and returns a determined value.
+    ///
+    /// - returns: The determined value, if already filled, or `nil`.
+    func peek() -> Value? {
+        return wait(.Now)
+    }
+
+    /// Waits for the value to become determined, then returns it.
+    ///
+    /// This is equivalent to unwrapping the value of calling `wait(.Forever)`,
+    /// but may be more efficient.
+    ///
+    /// This getter will unnecessarily block execution. It might be useful for
+    /// testing, but otherwise it should be strictly avoided.
+    ///
+    /// - returns: The determined value.
+    internal var value: Value {
+        return unsafeUnwrap(wait(.Forever))
+    }
+}
+
+public extension FutureType {
+    /// Begins another asynchronous operation with the deferred value once it
+    /// becomes determined.
+    ///
+    /// `flatMap` is similar to `map`, but `transform` returns a `Deferred`
+    /// instead of an immediate value. Use `flatMap` when you want this future
+    /// to feed into another asynchronous operation. You might hear this
+    /// referred to as "chaining" or "binding".
+    ///
+    /// - parameter queue: Optional dispatch queue for starting the new
+    ///   operation from. Defaults to a global queue matching the current QoS.
+    /// - parameter transform: Start a new operation using the deferred value.
+    /// - returns: The new deferred value returned by the `transform`.
+    /// - seealso: Deferred
+    func flatMap<NewFuture: FutureType>(upon queue: dispatch_queue_t = genericQueue, transform: Value -> NewFuture) -> Deferred<NewFuture.Value> {
+        let d = Deferred<NewFuture.Value>()
+        upon(queue) {
+            transform($0).upon(queue) {
+                d.fill($0, assertIfFilled: true)
```

Comment:
Yes, this has been under discussion for a while. I advocated fill not exploding due to how I understand the semantics of `Deferred`. Removing an independent `Deferred` in favor of a `Future` (read) and `Promise` (write) pair routes around that question and is a lot clearer, at the cost of having to keep track of two things rather than just one in client code.